### DISCUSSION TITLE
Canonical implementation of very simple shared state

### DISCUSSION THREAD

**Discussion by emiruz** [tokio-rs/axum#1756] [General]
I'm trying to implement shared state (to cache expensive functions in this instance) as an Arc wrapped Mutex. I assume others will find this useful too. Here is a minimal pseudo-code example for how I'm trying to do it.  I'd be grateful if you could guide me on whether (1) this approach makes sense, and (2) if there is a more canonical way?

        use std::sync::{Arc, Mutex};
        
        pub struct CacheState {
           pub cache: Vec<String>
        }
        
        type SharedState = State<Arc<Mutex<CacheState>>>;
        
        
        #[tokio::main]
        async fn main() {
        
            let state = Mutex::new(CacheState{cache: Vec::new()});
            let state =	Arc::new(state);
            
            let app =
	        Router::new()
	        .route("/:param/", get(example))
	        .with_state(state);
            
            axum::Server::bind(&"0.0.0.0:3000".parse().unwrap())
                .serve(app.into_make_service())
                .await
                .unwrap();
        }
        
        
        async fn example(Path((param)): Path(String), State(state): SharedState) -> impl IntoResponse {
            
            // i assume i dont need to call state.clone() because
            // it will implicitly happen?
        
            if let Ok(mut x) = self.lock() {
                x.cache.push("Put something in the vector".to_string());
            }
        }

**Comment by davidpdrsn** [comment]
Caching is a huge topic so hard to say give a "one true way" to do it. Among other things it depends on

- Do you care about cache invalidation?
- Is the operation you're caching async?
- Do you need a cache thats shared among server instances?

But if all you want is the simplest possible thing, and the operation to cache doesn't do IO, then what you posted is a very common way of doing it.

**Reply by davidpdrsn** [reply]
> i assume i dont need to call state.clone() because

Yes. If it didn't happen would would axum be able to give the function an owned `SharedState`? Otherwise it would have to give a `&SharedState`, which isn't possible because reasons 😜

**Reply by emiruz** [reply]
Thank you :) I'm a Rust noob, so thought it best that I ask.

**Comment by emiruz** [comment]
Thank you @davidpdrsn for the response. It is everything that I need to know. I've changed the title to include "very simple" so that it may be more useful to others.
